src/tigerbeetle/scrub.zig

```diff
@@ -0,0 +1,510 @@
+        &scrub,
+        gpa,
+        io,
+        tracer,
+        args.path,
+        args.lsm_forest_node_count,
+    );
+    defer scrub.deinit(gpa);
+
+    scrubbed_bytes += scrubbed_bytes_superblock;
+
+    try scrub.open();
+
+    if (!args.skip_wal) {
+        const scrubbed_bytes_wal = try scrub.scrub_wal();
+        assert(
+            scrubbed_bytes_wal == vsr.Zone.wal_headers.size().? + vsr.Zone.wal_prepares.size().?,
+        );
+
+        scrubbed_bytes += scrubbed_bytes_wal;
+    }
+
+    if (!args.skip_client_replies) {
+        const scrubbed_bytes_client_replies = try scrub.scrub_client_replies();
+        assert(scrubbed_bytes_client_replies == vsr.Zone.client_replies.size().?);
+
+        scrubbed_bytes += scrubbed_bytes_client_replies;
+    }
+
+    const grid_blocks_expected_count =
+        (scrub.grid.free_set.count_acquired() - scrub.grid.free_set.count_released()) +
+        scrub.grid.free_set_checkpoint_blocks_acquired.block_count() +
+        scrub.grid.free_set_checkpoint_blocks_released.block_count();
+
+    if (!args.skip_grid) {
+        // If no seed was given, use a random seed for better coverage.
+        const seed: u64 = seed_from_arg: {
+            const seed_argument = args.seed orelse
+                break :seed_from_arg @truncate(stdx.unique_u128());
+            break :seed_from_arg vsr.testing.parse_seed(seed_argument);
+        };
+
+        const scrubbed_bytes_grid = try scrub.scrub_grid(seed);
+        assert(scrubbed_bytes_grid == grid_blocks_expected_count * constants.block_size);
+
+        scrubbed_bytes += scrubbed_bytes_grid;
+    }
+
+    const scrubbed_bytes_target: u64 = blk: {
+        var scrubbed_bytes_target: u64 = 0;
+
+        for (std.enums.values(vsr.Zone)) |zone| {
+            if (vsr.Zone.size(zone)) |size| {
+                scrubbed_bytes_target += size;
+            }
+        }
+        scrubbed_bytes_target += grid_blocks_expected_count * constants.block_size;
+        scrubbed_bytes_target -= vsr.Zone.grid_padding.size().?;
+
+        // At this point, scrubbed_bytes_target is the highest it will ever be if everything were
+        // scrubbed. Then, subtract off the skipped items explicitly.
+
+        if (args.skip_wal) {
+            scrubbed_bytes_target -= vsr.Zone.wal_headers.size().? + vsr.Zone.wal_prepares.size().?;
+        }
+
+        if (args.skip_client_replies) {
+            scrubbed_bytes_target -= vsr.Zone.client_replies.size().?;
+        }
+
+        if (args.skip_grid) {
+            scrubbed_bytes_target -= grid_blocks_expected_count * constants.block_size;
+        }
+
+        break :blk scrubbed_bytes_target;
+    };
+
+    assert(scrubbed_bytes == scrubbed_bytes_target);
+}
+
+const Scrub = @This();
+
+io: *IO,
+storage: Storage,
+
+superblock: SuperBlock,
+client_sessions_checkpoint: CheckpointTrailer,
+grid: Grid,
+forest: Forest,
+grid_scrubber: GridScrubber,
+grid_blocks_scrubbed: std.AutoArrayHashMapUnmanaged(u64, void),
```

Comment:
It seems like this could be a bitset instead? Though hash map is fine too, whichever you prefer.